### PR TITLE
XAML Jest Snapshots

### PR DESCRIPTION
**Pull request by NickGerleman**
This PR removes the tree-dump control, allowing the dumping of arbitrary visual content from the test runner. This is done by doing RPC from the test runner to test app. This RPC channel can be easily extended in the future to allow running more arbitrary native commands from test runner.

Diffing is now handled by the industry standard Jest snapshots, instead of custom logic in TreeDumpControl to handle assertions. We replicate tree dump comparison logic of removing collapsed subtrees, and dumps themselves omit scrollbars. 

Dump failures are shown inline in the test runner and in CI.
image

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7118)

### CODE REVIEW

File: packages/e2e-test-app/windows/ReactUWPTestApp/MainPage.xaml.cs [hunk 1]
```diff
@@ -16,6 +23,42 @@ public MainPage()
...
+
+
+            var handler = new Handler();
+            handler.BindOperation("DumpVisualTree", DumpVisualTree);
+
+            var server = new Server(handler);
+            var tsk = LoopServer(server);
         }
+
+        JsonObject DumpVisualTree(JsonValue payload)
```
Comment (NickGerleman) [inline packages/e2e-test-app/windows/ReactUWPTestApp/MainPage.xaml.cs:35]:
```suggestion
        JsonObject DumpVisualTree(JsonValue parameters)
```

**Comment by NickGerleman** [comment]
Think CI should be good after this one. https://github.com/microsoft/react-native-windows/pull/7117 confirmed artifacts are being correctly uploaded on test failure.

**Comment by NickGerleman** [comment]
Looks like this still needs measurement fixes. Need to figure out the right way to create insensitive dump for sensitive compare.

**Comment by NickGerleman** [comment]
Not sure if normalizing before exact comparison can work. Need to think more about the scenario.

File: packages/e2e-test-app/test/framework/TreeDump.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * Round measurements, to account for differences in scale-factor, rounding,
+ * etc.
+ */
+function roundMeasurements(element: UIElement) {
+  // Allow measuresments to be 2.5 DIPs off
+  const DIPS_FUZZINESS = 2.5;
+
+  const roundFactor = DIPS_FUZZINESS * 2;
```
Comment (NickGerleman) [inline packages/e2e-test-app/test/framework/TreeDump.ts:99]: This isn't right at all. Still need to think on whether there is a way to achieve what I'm looking for.

**Comment by NickGerleman** [comment]
Should add eslint ban on toMatchSnapshot to force usage of fuzzy compare version instead.

**Review by dannyvv — APPROVED** [review]

**Comment by NickGerleman** [comment]
Hmm, the serialized property matchers are also compared... Need to think harder on whether there's a way to add fuzzy matching here.

File: .ado/jobs/e2e-test.yml [hunk 3]
```diff
@@ -114,11 +100,11 @@ jobs:
...
       - task: PublishPipelineArtifact@1
-        displayName: "Publish Artifact:ReactUWPTestAppTreeDump"
+        displayName: "Publish Artifact:Snapshots"
         inputs:
-          artifactName: ReactUWPTestAppTreeDump-$(BuildPlatform)-$(System.JobAttempt)
-          targetPath: $(Build.StagingDirectory)/ReactUWPTestAppTreeDump
-        condition: succeededOrFailed()
+          artifactName: Snapshots-$(BuildPlatform)-$(System.JobAttempt)
+          targetPath: $(Build.StagingDirectory)/snapshots
+        condition: failed()
```
[the remark this answers is not in the document]
Comment (NickGerleman) [reply .ado/jobs/e2e-test.yml:107]: The diff will be shown in the initial failure:  
image

Having a direct link to snapshot artifacts could be nice though. The artifacts UI is not intuitive to newcomers.

**Comment by NickGerleman** [comment]
Okay, this is all passing now for real this time.

Tried various workarounds to allow fuziness in RenderSize but nothing works well with snapshots. RenderSize is an output of other deterministic inputs in the tree though, so I think we're still fine. E.g. Yoga sets raw width and height, or bits that are self measure have inputs like text.

**Pull request merged**